CMakeLists.txt

```diff
@@ -679,18 +679,8 @@ if(BUILD_CPP_TEST)
   file(GLOB CPP_TEST_SOURCES tests/cpp_tests/*.cpp)
   if(MSVC)
     set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /permissive-")
-    set(
-      CompilerFlags
-        CMAKE_CXX_FLAGS
-        CMAKE_CXX_FLAGS_DEBUG
-        CMAKE_CXX_FLAGS_RELEASE
-        CMAKE_C_FLAGS
-        CMAKE_C_FLAGS_DEBUG
-        CMAKE_C_FLAGS_RELEASE
-    )
-    foreach(CompilerFlag ${CompilerFlags})
-      string(REPLACE "/MD" "/MT" ${CompilerFlag} "${${CompilerFlag}}")
-    endforeach()
+    # Use /MT flag to statically link the C runtime
+    set(CMAKE_MSVC_RUNTIME_LIBRARY "MultiThreaded$<$<CONFIG:Debug>:Debug>")
```

Comment:
@hcho3 could you please look into https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=16422&view=logs&j=1f4df553-f999-5fff-c6fe-71123c872ae2&t=3537c7a9-e1e2-54d5-18f4-f7d9e02b9f6d?

```text
boosting.obj : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MDd_DynamicDebug' doesn't match value 'MTd_StaticDebug' in test_array_args.obj [D:\a\1\s\build\testlightgbm.vcxproj]
```

Does this maybe need to be set earlier, before the target is declared, to affect the flags? I'm not really sure.

Reply:
Yes, I think this has to be set befor3 targets are created.